### PR TITLE
add tests for copp manager redesign

### PR DESCRIPTION
**Pull request by JibinBao**
### Description of PR

Add tests  for CoPP Manager redesign feature
1. CoPP Manger redesign test plan: https://github.com/Azure/SONiC/pull/903
2. HLD: https://github.com/noaOrMlnx/sonic-swss/blob/4865a8eb37444c2c070ef49cc69690e4295ce07d/doc/copp_redisgn.md

Summary:
Fixes # (issue)

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

### Back port request
- [ ] 201911

#### What is the motivation for this PR?
Add new tests for CoPP Redesign feature.

#### How did you do it?
Add new tests

#### How did you verify/test it?
Run New CoPP tests

### CODE REVIEW

**Review by nhe-NV — COMMENTED** [review]
It should be bool

File: tests/copp/copp_utils.py [hunk 1]
```diff
@@ -236,3 +243,128 @@ def _get_http_and_https_proxy_ip(creds):
...
+def install_trap(dut, trap):
+    """
+    Install trap by setting always_enable to True
+
+    Args:
+        dut (SonicHost): The target device
+        feature_name (str): feature name corresponding to the trap
+        trap (str): trap name
+    """
+    configure_always_enabled_for_trap(dut, trap, "true")
```
Comment (nhe-NV) [inline tests/copp/copp_utils.py]: the value should be bool as described in the function define

File: tests/copp/copp_utils.py [hunk 2]
```diff
@@ -236,3 +243,128 @@ def _get_http_and_https_proxy_ip(creds):
...
+
+
+def install_trap(dut, trap):
+    """
+    Install trap by setting always_enable to True
+
+    Args:
+        dut (SonicHost): The target device
+        feature_name (str): feature name corresponding to the trap
+        trap (str): trap name
```
Comment (nhe-NV) [inline tests/copp/copp_utils.py]: this should be trap name or trap id? in the configure_always_enabled_for_trap it is described as trap id

File: tests/copp/copp_utils.py [hunk 3]
```diff
@@ -236,3 +243,128 @@ def _get_http_and_https_proxy_ip(creds):
 
     return (re.findall(r'[0-9]+(?:\.[0-9]+){3}', creds.get('proxy_env', {}).get('http_proxy', ''))[0],
             re.findall(r'[0-9]+(?:\.[0-9]+){3}', creds.get('proxy_env', {}).get('https_proxy', ''))[0])
+
+
+def configure_always_enabled_for_trap(dut, trap_id, always_enabled):
+    """
+    Configure the always_enabled to True or False for the specified trap Id.
```
Comment (nhe-NV) [inline tests/copp/copp_utils.py]: trap_id： the explain if trap id should be consistent

File: tests/copp/copp_utils.py [hunk 4]
```diff
@@ -236,3 +243,128 @@ def _get_http_and_https_proxy_ip(creds):
...
+    dut.command("sudo config load {} -y".format(copp_trap_config_json))
+
+
+def get_config_db_json_obj(dut):
+    """
+    Get config_db content from dut
+    Args:
+        dut (SonicHost): The target device
+    """
+    config_db_json = dut.shell("/usr/local/bin/sonic-cfggen -d --print-data")["stdout"]
```
Comment (nhe-NV) [inline tests/copp/copp_utils.py]: sudo sonic-cfggen

File: tests/copp/copp_utils.py [hunk 5]
```diff
@@ -236,3 +243,128 @@ def _get_http_and_https_proxy_ip(creds):
...
             re.findall(r'[0-9]+(?:\.[0-9]+){3}', creds.get('proxy_env', {}).get('https_proxy', ''))[0])
+
+
+def configure_always_enabled_for_trap(dut, trap_id, always_enabled):
+    """
+    Configure the always_enabled to True or False for the specified trap Id.
+    Args:
+        dut (SonicHost): The target device
+        trap_id (str): The trap ID (e.g. bgp)
+        always_enabled (Bool): True or False
```
Comment (nhe-NV) [inline tests/copp/copp_utils.py]: should be string, not bool

File: tests/copp/copp_utils.py [hunk 6]
```diff
@@ -236,3 +243,128 @@ def _get_http_and_https_proxy_ip(creds):
...
+
+
+def uninstall_trap(dut, feature_name, trap):
+    """
+    Uninstall trap by disabling feature and set always_enable to false
+
+    Args:
+        dut (SonicHost): The target device
+        feature_name (str): feature name corresponding to the trap
+        trap (str): trap name
```
Comment (nhe-NV) [inline tests/copp/copp_utils.py]: is it trap id?

File: tests/copp/copp_utils.py [hunk 7]
```diff
@@ -236,3 +243,128 @@ def _get_http_and_https_proxy_ip(creds):
...
+
+
+def verify_always_enable_value(dut, trap, always_enable_value):
+    """
+    Uninstall trap by disabling feature and set always_enable to False
+
+    Args:
+        dut (SonicHost): The target device
+        always_enable_value (bool): True or False
+        trap (str): trap name
```
Comment (nhe-NV) [inline tests/copp/copp_utils.py]: trap id or trap name?

File: tests/copp/copp_utils.py [hunk 8]
```diff
@@ -236,3 +243,128 @@ def _get_http_and_https_proxy_ip(creds):
...
+    configure_always_enabled_for_trap(dut, trap, "false")
+
+
+def verify_always_enable_value(dut, trap, always_enable_value):
+    """
+    Uninstall trap by disabling feature and set always_enable to False
+
+    Args:
+        dut (SonicHost): The target device
+        always_enable_value (bool): True or False
```
Comment (nhe-NV) [inline tests/copp/copp_utils.py]: string or bool?

File: tests/copp/test_copp.py [hunk 9]
```diff
@@ -55,11 +60,15 @@
...
 _TEST_RATE_LIMIT = 600
 
+logger = logging.getLogger(__name__)
+
 
 class TestCOPP(object):
     """
         Tests basic COPP functionality in SONiC.
     """
+    trap_name = "bgp"
```
Comment (nhe-NV) [inline tests/copp/test_copp.py]: need to make it consistent for trap name , trap id

File: tests/copp/test_copp.py [hunk 10]
```diff
@@ -98,6 +107,122 @@ def test_no_policer(self, protocol, duthosts, enum_rand_one_per_hwsku_frontend_h
                      copp_testbed,
                      dut_type)
 
+    def test_add_new_trap(self, duthosts, enum_rand_one_per_hwsku_frontend_hostname, ptfhost, check_image_version, copp_testbed, dut_type, backup_restore_config_db):
+        """
+        Validates that one new tarp(bgp) can be installed
```
Comment (nhe-NV) [inline tests/copp/test_copp.py]: typo

File: tests/copp/test_copp.py [hunk 11]
```diff
@@ -98,6 +107,122 @@ def test_no_policer(self, protocol, duthosts, enum_rand_one_per_hwsku_frontend_h
...
+        copp_utils.configure_always_enabled_for_trap(duthost, self.trap_name, "true")
+
+        logger.info("Verify {} trap status is installed by sending traffic".format(self.trap_name))
+        pytest_assert(
+            wait_until(60, 20, 0, _copp_runner, duthost, ptfhost, self.trap_name.upper(), copp_testbed, dut_type),
+            "Installing {} trap fail".format(self.trap_name))
+
+    def test_remove_trap_by_deleting_feature_entry(self, duthosts, enum_rand_one_per_hwsku_frontend_hostname, ptfhost, check_image_version, copp_testbed, dut_type, backup_restore_config_db):
+        """
+        Validates that The tarp(bgp) can be uninstalled after deleting the corresponding entry from the feature table
```
Comment (nhe-NV) [inline tests/copp/test_copp.py]: typo

Comment (nhe-NV) [reply tests/copp/test_copp.py]: the not The

File: tests/copp/test_copp.py [hunk 12]
```diff
@@ -98,6 +107,122 @@ def test_no_policer(self, protocol, duthosts, enum_rand_one_per_hwsku_frontend_h
...
+
+        1. Set always_enabled of bgp to True to make sure trap(bgp) is installed
+        2. Remove bgp entry from feature table with CMD follow:
```
Comment (nhe-NV) [inline tests/copp/test_copp.py]: True to true

File: tests/copp/test_copp.py [hunk 13]
```diff
@@ -98,6 +107,122 @@ def test_no_policer(self, protocol, duthosts, enum_rand_one_per_hwsku_frontend_h
...
+           redis-cli -n 4 hdel "FEATURE|bgp" "auto_restart" "has_global_scope" "has_per_asic_scope" "has_timer" "high_mem_alert" "state"
+        3. Verify the trap status is uninstalled by sending traffic
+        """
+        duthost = duthosts[enum_rand_one_per_hwsku_frontend_hostname]
+
+        logger.info("Pre condition: make trap {} is installed".format(self.trap_name))
+        pre_condition_install_trap(ptfhost, duthost, copp_testbed, self.trap_name)
+
+        logger.info("Set always_enabled of {} to False".format(self.trap_name))
```
Comment (nhe-NV) [inline tests/copp/test_copp.py]: false

File: tests/copp/test_copp.py [hunk 14]
```diff
@@ -98,6 +107,122 @@ def test_no_policer(self, protocol, duthosts, enum_rand_one_per_hwsku_frontend_h
...
+        copp_utils.remove_feature_entry(duthost, self.feature_name)
+
+        logger.info("Verify {} trap status is uninstalled by sending traffic".format(self.trap_name))
+        pytest_assert(
+            wait_until(100, 20, 0, _copp_runner, duthost, ptfhost, self.trap_name.upper(), copp_testbed, dut_type, has_trap=False),
+            "uninstalling {} trap fail".format(self.trap_name))
+
+    def test_remove_trap_by_disabling_feature(self, duthosts, enum_rand_one_per_hwsku_frontend_hostname, check_image_version, ptfhost, copp_testbed, dut_type, backup_restore_config_db):
+        """
+        Validates that The tarp(bgp) can be uninstalled after disabling the corresponding entry from the feature table
```
Comment (nhe-NV) [inline tests/copp/test_copp.py]: typo

File: tests/copp/test_copp.py [hunk 15]
```diff
@@ -98,6 +107,122 @@ def test_no_policer(self, protocol, duthosts, enum_rand_one_per_hwsku_frontend_h
...
+
+        logger.info("Verify {} trap status is uninstalled by sending traffic".format(self.trap_name))
+        pytest_assert(
+            wait_until(100, 20, 0, _copp_runner, duthost, ptfhost, self.trap_name.upper(), copp_testbed, dut_type,
+                       has_trap=False),
+            "Uninstalling {} trap fail".format(self.trap_name))
+
+    def test_trap_config_save_after_reboot(self, duthosts, localhost, enum_rand_one_per_hwsku_frontend_hostname, ptfhost,check_image_version, copp_testbed, dut_type, backup_restore_config_db):
+        """
+        Validates that The tarp configuration is saved or not after reboot(reboot, fast-reboot, warm-reboot)
```
Comment (nhe-NV) [inline tests/copp/test_copp.py]: the not The

File: tests/copp/test_copp.py [hunk 16]
```diff
@@ -98,6 +107,122 @@ def test_no_policer(self, protocol, duthosts, enum_rand_one_per_hwsku_frontend_h
...
+
+        1. Set always_enabled of a trap(e.g. bgp) to True
```
Comment (nhe-NV) [inline tests/copp/test_copp.py]: true not True

File: tests/copp/copp_utils.py [hunk 17]
```diff
@@ -236,3 +243,127 @@ def _get_http_and_https_proxy_ip(creds):
...
+    dut.command("sudo cp {} {}".format(_CONFIG_DB, _TEMP_CONFIG_DB))
+
+
+def restore_config_db(dut):
+    """
+    Restore config db
+    Args:
+        dut (SonicHost): The target device
+    """
+    dut.command("sudo cp {} {}".format(_TEMP_CONFIG_DB, _CONFIG_DB))
```
Comment (nhe-NV) [inline tests/copp/copp_utils.py:328]: should delete the tmp_config_db?

**Review by nhe-NV — APPROVED** [review]

File: tests/copp/test_copp.py [hunk 18]
```diff
@@ -331,3 +461,34 @@ def _teardown_multi_asic_proxy(dut, creds, test_params, tbinfo):
...
+    logger.info("Verify {} trap status is installed by sending traffic in pre_condition".format(trap_id))
+
+    pytest_assert(
+        wait_until(100, 20, 0, _copp_runner, duthost, ptfhost, trap_id.upper(), copp_testbed, dut_type),
+        "Installing {} trap fail".format(trap_id))
+
+
+@pytest.fixture(autouse=False, scope="class")
+def check_image_version(duthosts, enum_rand_one_per_hwsku_frontend_hostname):
+    """Skips this test if the SONiC image installed on DUT is older than 202106
```
Comment (noaOrMlnx) [inline tests/copp/test_copp.py]: copp new logic should be in 202012 branch and above

File: tests/copp/test_copp.py [hunk 19]
```diff
@@ -331,3 +461,34 @@ def _teardown_multi_asic_proxy(dut, creds, test_params, tbinfo):
...
+
+    Args:
+        duthost: Hostname of DUT.
+
+    Returns:
+        None.
+    """
+    skip_release(duthosts[enum_rand_one_per_hwsku_frontend_hostname], ["201911"])
```
Comment (noaOrMlnx) [inline tests/copp/test_copp.py]: what about older releases? 

Also, please add new line at EOF

File: tests/copp/test_copp.py [hunk 20]
```diff
@@ -98,6 +107,123 @@ def test_no_policer(self, protocol, duthosts, enum_rand_one_per_hwsku_frontend_h
...
+        logger.info("Verify {} trap status is installed by sending traffic".format(self.trap_id))
+        pytest_assert(
+            wait_until(60, 20, 0, _copp_runner, duthost, ptfhost, self.trap_id.upper(), copp_testbed, dut_type),
+            "Installing {} trap fail".format(self.trap_id))
+
+    def test_remove_trap_by_deleting_feature_entry(self, duthosts, enum_rand_one_per_hwsku_frontend_hostname, ptfhost, check_image_version, copp_testbed, dut_type, backup_restore_config_db):
+        """
+        Validates that The trap(bgp) can be uninstalled after deleting the corresponding entry from the feature table
+
+        1. Set always_enabled of bgp to true to make sure trap(bgp) is installed
```
Comment (noaOrMlnx) [inline tests/copp/test_copp.py]: why would you set always_enable to true and then delete the feature entry? 
if always_enabled is true, and feature entry is removed, the trap should be installed.

File: tests/copp/copp_utils.py [hunk 21]
```diff
@@ -236,3 +243,128 @@ def _get_http_and_https_proxy_ip(creds):
...
+        feature_name (str): feature name corresponding to the trap
+        trap_id (str): trap id
+    """
+    disable_feature_entry(dut, feature_name)
+    configure_always_enabled_for_trap(dut, trap_id, "false")
+
+
+def verify_always_enable_value(dut, trap_id, always_enable_value):
+    """
+    Uninstall trap by disabling feature and set always_enable to false
```
Comment (noaOrMlnx) [inline tests/copp/copp_utils.py]: please fix the comment and args

File: tests/copp/copp_utils.py [hunk 22]
```diff
@@ -236,3 +243,128 @@ def _get_http_and_https_proxy_ip(creds):
...
+
+
+def remove_feature_entry(dut, feature_name):
+    """
+    Remove feature entry from dut
+    Args:
+        dut (SonicHost): The target device
+        feature_name (str): feature name (e.g bgp)
+    """
+    dut.command('redis-cli -n 4 hdel "FEATURE|{}" "auto_restart" "has_global_scope" "has_per_asic_scope" "has_timer" '
```
Comment (noaOrMlnx) [inline tests/copp/copp_utils.py]: can be changed to "redis-cli -n 4 del FEATURE|name"
it will delete the whole entry

**Review by noaOrMlnx — CHANGES_REQUESTED** [review]

File: tests/copp/copp_utils.py (hunk 21 above)
Comment (JibinBao) [reply tests/copp/copp_utils.py]: fix

File: tests/copp/test_copp.py (hunk 20 above)
Comment (JibinBao) [reply tests/copp/test_copp.py]: Actually I set the always_enable to false later.
I think you suggestion is better, I will install the trap by enabling feature status

File: tests/copp/copp_utils.py (hunk 22 above)
Comment (JibinBao) [reply tests/copp/copp_utils.py]: fix

File: tests/copp/test_copp.py (hunk 18 above)
Comment (JibinBao) [reply tests/copp/test_copp.py]: ok

File: tests/copp/test_copp.py (hunk 19 above)
Comment (JibinBao) [reply tests/copp/test_copp.py]: For 201811 we don't support it anymore

**Review by noaOrMlnx — APPROVED** [review]

**Comment by JibinBao** [comment]
close it, and open external one

**Pull request closed without merging**